### PR TITLE
curvefs/client: add partition relative function for metacache

### PR DESCRIPTION
**Pull request by ilixiaocui**
### What problem does this PR solve?

Issue Number: close #xxx 

Problem Summary:

### What is changed and how it works?

What's Changed:

How it Works:

Side effects(Breaking backward compatibility? Performance regression?): 

### Check List

- [ ] Relevant documentation/comments is changed or added
- [ ] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

**Comment by ilixiaocui** [comment]
recheck

File: curvefs/src/client/rpcclient/metacache.cpp [hunk 1]
```diff
@@ -183,6 +189,176 @@ bool MetaCache::IsLeaderMayChange(const CopysetGroupID &groupID) {
...
+        return false;
+    }
+
+    // add partition and copyset info
+    WriteLockGuard wl4PartitionMap(rwlock4Partitions_);
+    WriteLockGuard wl4CopysetMap(rwlock4copysetInfoMap_);
+    DoAddPartitionAndCopyset(partitionInfos, copysetMap);
+}
+
+bool MetaCache::DoListOrCreatePartitions(
```
Comment (wu-hanqing) [inline curvefs/src/client/rpcclient/metacache.cpp:256]: so many RPC interactive, can we just send one RPC and get what we want?

File: curvefs/src/client/rpcclient/metacache.cpp [hunk 2]
```diff
@@ -217,56 +393,96 @@ bool MetaCache::UpdateLeaderInternal(
...
-            if (!find && pinfo.start() <= inodeID && pinfo.end() >= inodeID) {
-                find = true;
-                *groupID = std::move(
-                    CopysetGroupID(pinfo.poolid(), pinfo.copysetid()));
-                *partitionID = pinfo.partitionid();
-            }
-        });
+    for_each(partitionInfos_.begin(), partitionInfos_.end(),
+             [&](const PartitionInfo &pinfo) {
+                 if (!find && pinfo.start() <= inodeID &&
```
Comment (wu-hanqing) [inline curvefs/src/client/rpcclient/metacache.cpp:472]: why not directly return when found ?

File: curvefs/src/client/rpcclient/task_excutor.cpp [hunk 3]
```diff
@@ -77,6 +77,11 @@ bool TaskExecutor::OnReturn(int retCode) {
             OnCopysetNotExist();
             break;
 
+        case MetaStatusCode::PARTITION_ALLOC_ID_FAIL:
+            // need choose a new coopyset
```
Comment (wu-hanqing) [inline curvefs/src/client/rpcclient/task_excutor.cpp:81]: missing needRetry = true ?

File: curvefs/src/client/rpcclient/task_excutor.cpp [hunk 4]
```diff
@@ -176,6 +181,9 @@ void TaskExecutor::RefreshLeader() {
     task_->retryDirectly = (oldTarget != task_->target.metaServerID);
 }
 
+void TaskExecutor::OnPartitionAllocIDFail() {
+    metaCache_->MarkPartitionFull(task_->target.partitionID);
```
Comment (wu-hanqing) [inline curvefs/src/client/rpcclient/task_excutor.cpp:185]: should we let the full state persists to MDS ?
e.g., if we have a huge filesystem that contains hundreds of full partitions, and create inode will fail hundreds of times, and in the last, we create a new partition.

another possible solution is if create inode got three consecutive failures by `PARTITION_ALLOC_ID_FAIL`, create a new partition immediately

File: curvefs/src/client/rpcclient/mds_client.cpp [hunk 5]
```diff
@@ -316,6 +317,22 @@ bool MdsClientImpl::GetMetaServerListInCopysets(
...
+    const std::vector<uint32_t> &partitionIDList,
+    std::map<uint32_t, Copyset> *copysetMap) {
+    return true;
+}
+
+bool MdsClientImpl::ListPartition(uint32_t fsID,
+                                  std::vector<PartitionInfo> *partiitonInfos) {
+    return true;
+}
+
```
[the remark this answers is not in the document]
Comment (ilixiaocui) [reply curvefs/src/client/rpcclient/mds_client.cpp:335]: here @cw123 will implement.

File: curvefs/src/client/rpcclient/metacache.h [hunk 6]
```diff
@@ -42,11 +43,12 @@ using ::curve::client::CopysetID;
 using ::curve::client::CopysetInfo;
 using ::curve::client::LogicPoolID;
 using ::curve::common::RWLock;
+using ::curvefs::client::common::MetaCacheOpt;
 using ::curvefs::client::common::MetaserverID;
 using ::curvefs::client::common::MetaServerOpType;
+using ::curvefs::client::common::Mutex;
```
[the remark this answers is not in the document]
Comment (ilixiaocui) [reply curvefs/src/client/rpcclient/metacache.h:49]: Mutex is used with std::lock_guard, not RWLock

File: curvefs/src/client/rpcclient/metacache.cpp [hunk 7]
```diff
@@ -183,6 +189,176 @@ bool MetaCache::IsLeaderMayChange(const CopysetGroupID &groupID) {
...
+            LOG(ERROR) << "create partition for fsid:" << fsID_ << " fail";
+            return false;
+        }
+    }
+
+    // gather partitionIDs
+    std::vector<PartitionID> partitionIDLsit;
+    for_each(partitionInfos->begin(), partitionInfos->end(),
+             [&](const PartitionInfo &info) {
+                 auto key = std::move(CalcLogicPoolCopysetID(
```
[the remark this answers is not in the document]
Comment (ilixiaocui) [reply curvefs/src/client/rpcclient/metacache.cpp:276]: fix

File: curvefs/src/client/rpcclient/metacache.cpp (hunk 1 above)
Comment (ilixiaocui) [reply curvefs/src/client/rpcclient/metacache.cpp:256]: yes, need to be modify. I mark todo here

File: curvefs/src/client/rpcclient/metacache.cpp [hunk 8]
```diff
@@ -183,6 +189,176 @@ bool MetaCache::IsLeaderMayChange(const CopysetGroupID &groupID) {
...
+                     CopysetGroupID(info.poolid(), info.copysetid())));
+                 bool exist = false;
+
+                 if (!list) {
+                     ReadLockGuard rl(rwlock4copysetInfoMap_);
+                     exist = (copysetInfoMap_.count(key) > 0);
+                 }
+
+                 if (!exist) {
+                     partitionIDLsit.emplace_back(info.partitionid());
```
Comment (xu-chaojie) [inline curvefs/src/client/rpcclient/metacache.cpp:286]: if exist copyset in copysetInfoMap_, then  mdsClient_->CreatePartition has no effect?

File: curvefs/src/client/rpcclient/metacache.cpp (hunk 2 above)
Comment (ilixiaocui) [reply curvefs/src/client/rpcclient/metacache.cpp:472]: here for_each is inappropriate

File: curvefs/src/client/rpcclient/task_excutor.cpp (hunk 3 above)
Comment (ilixiaocui) [reply curvefs/src/client/rpcclient/task_excutor.cpp:81]: fix

File: curvefs/src/client/rpcclient/metacache.cpp [hunk 9]
```diff
@@ -183,6 +189,176 @@ bool MetaCache::IsLeaderMayChange(const CopysetGroupID &groupID) {
...
+    } else {
+        if (!mdsClient_->CreatePartition(
+                fsID_, metacacheopt_.createPartitionOnce, partitionInfos)) {
+            LOG(ERROR) << "create partition for fsid:" << fsID_ << " fail";
+            return false;
+        }
+    }
+
+    // gather partitionIDs
+    std::vector<PartitionID> partitionIDLsit;
```
Comment (xu-chaojie) [inline curvefs/src/client/rpcclient/metacache.cpp:273]: partitionIDLsit -> partitionIDList

File: curvefs/src/client/rpcclient/task_excutor.cpp (hunk 4 above)
Comment (ilixiaocui) [reply curvefs/src/client/rpcclient/task_excutor.cpp:185]: yes, I have discuss with @cw123 ,  heartbeat with metaserver and mds should report this message. I mark TODO here

File: curvefs/src/client/rpcclient/metacache.cpp [hunk 10]
```diff
@@ -183,6 +188,178 @@ bool MetaCache::IsLeaderMayChange(const CopysetGroupID &groupID) {
...
+                     exist = (copysetInfoMap_.count(key) > 0);
+                 }
+
+                 if (!exist) {
+                     partitionIDList.emplace_back(info.partitionid());
+                 }
+             });
+
+    // get copyset for each partition
+    std::map<PartitionID, Copyset> copysets;
```
Comment (xu-chaojie) [inline curvefs/src/client/rpcclient/metacache.cpp:292]: if partitionIDList can be empty here, add a if clause here

**Review by wu-hanqing — APPROVED** [review]

File: curvefs/src/client/rpcclient/task_excutor.cpp [hunk 11]
```diff
@@ -77,6 +77,14 @@ bool TaskExecutor::OnReturn(int retCode) {
             OnCopysetNotExist();
             break;
 
+        case MetaStatusCode::PARTITION_ALLOC_ID_FAIL:
```
Comment (xu-chaojie) [inline curvefs/src/client/rpcclient/task_excutor.cpp:80]: PARTITION_ALLOC_ID_FAIL. -> PARTITION_FULL
or
MarkPartitionFull ->  MarkPartitionUnavailable
below

File: curvefs/src/client/rpcclient/metacache.cpp [hunk 12]
```diff
@@ -292,23 +421,74 @@ void MetaCache::UpdateCopysetInfoIfMatchCurrentLeader(
     }
 }
 
-bool MetaCache::UpdateCopysetInfoFromMDS(
-    const CopysetGroupID &groupID, CopysetInfo<MetaserverID> *targetInfo) {
-    std::vector<CopysetInfo<MetaserverID>> metaServerInfos;
+bool MetaCache::SelectPartition(CopysetTarget *target) {
```
Comment (xu-chaojie) [inline curvefs/src/client/rpcclient/metacache.cpp:424]: This function never be called

Comment (ilixiaocui) [reply curvefs/src/client/rpcclient/metacache.cpp:425]: fix

File: curvefs/src/client/rpcclient/task_excutor.cpp (hunk 11 above)
Comment (ilixiaocui) [reply curvefs/src/client/rpcclient/task_excutor.cpp:80]: fix

**Comment by ilixiaocui** [comment]
recheck

File: curvefs/src/client/rpcclient/metacache.cpp [hunk 13]
```diff
@@ -183,6 +185,182 @@ bool MetaCache::IsLeaderMayChange(const CopysetGroupID &groupID) {
...
+
+    init_ = true;
+
+    LOG(INFO) << "init partition and copyset infos for {fsid:" << fsID
+              << "} ok";
+    return true;
+}
+
+bool MetaCache::CreatePartitions(int currentNum,
+                                 PatitionInfoList *newPartitions) {
```
Comment (xu-chaojie) [inline curvefs/src/client/rpcclient/metacache.cpp:228]: newPartitions is never assigned a value,  but is used outside.

**Review by xu-chaojie — APPROVED** [review]

**Pull request merged**